mdata/chunk/tsz/tszlong.go

```diff
@@ -106,11 +106,25 @@ func (s *SeriesLong) Push(t uint32, v float64) {
 			leading = 31
 		}
 
-		// TODO(dgryski): check if it's 'cheaper' to reset the leading/trailing bits instead
-		if s.leading != ^uint8(0) && leading >= s.leading && trailing >= s.trailing {
+		sigbits := 64 - leading - trailing
+		currbits := 64 - s.leading - s.trailing
+
+		// We have 2 cases here:
+		// 1. We have at least as many leading and trailing zeroes than we previously did
+		// 2. We have fewer leading or trailing zeroes than previously
+		// In the case of number 2, we need to re-encode the leading/trailing zeroes count to be able to handle the number
+		// of significant bits we need to write.
+		// In the case of number 1, we have more options. If we have significantly more leading/trailing zeroes, then we
+		// might be wasting a lot of bits by writing more significant bits than we need. In this case we *could* choose
+		// to "reset" by writing out the count of significant bits. It costs 11 bits to reset, so ideally we would only
+		// reset when we know we will waste more bits than it would cost to reset. Add another 11 bits in case this was
+		// an outlier and we need to 'reset' back to more significant bits.
```

Comment:
interesting heuristic.
You're basically only doing the reset if within this single push we get a "surplus benefit" of 11 bits. I wonder if this may lead to
1) subsequent writes being forced to reset again because subsequent writes may not "fit" anymore.
2) needlessly postponing beneficial resets.

I wonder if it may make sense to reset more aggressively, e.g. invest the 11bit reset overhead in return for, say a 5bit reduction in number of sigbits.  Even if the tradeoff for this single write isn't worth it, future writes would more likely also fit within the same sigbits and thus be able to enjoy the same benefit and we could amortize the reset cost over multiple writes.

At Grafana we don't need to squeeze out these last couple of percentage points from our chunk.  So up to you whether you want to test a bit more, or push this forward as-is.

Reply:
> I wonder if this may lead to subsequent writes being forced to reset again because subsequent writes may not "fit" anymore, lowering the average benefit.

I do believe that this happens, which is why I decided to not reset overly aggressively (hence the surplus 11). I had a working implementation that tracked the leading/trailing bits for the previous streams that was more complete but the additional tracking code was moderately complex and added fields tot he struct (which means memory usage). Even then, we are ultimately trying to predict the future from the past and it wasn't much more effective than this simple approach. 

> I wonder if this may lead to needlessly postponing beneficial resets.

Definitely, but the existing code never resets, so this is a "something is better than nothing" approach that is purposively a bit conservative.

FWIW, I had a thought that an "optimization" round could be performed on large chunks after they were complete but before they were stored. Since we don't need to predict the future at this point (we have all of the points already) we could likely effectively reduce chunks with a lot of resets. This wouldn't reduce memory usage but would reduce storage/network for writes and later reads. At this point in time it's not something we are looking to explore though.